virtual_rainforest/core/grid.py

```diff
@@ -1,203 +1,370 @@
+       data loading methods to assign values to grids? This would be a useful way of
+       defining mappings though.
 """
 
-import fiona
+import json
+import logging
+from functools import wraps
+from typing import Callable
+
 import numpy as np
-from pydantic import StrictInt
-from pydantic.dataclasses import dataclass
-from pydantic.typing import Literal
+from shapely.affinity import scale, translate
+from shapely.geometry import Polygon
 
+LOGGER = logging.getLogger("virtual_rainforest.core")
 
-@dataclass
-class CoreGridConfig:
-    """Configure the `core.grid` module.
+GRID_REGISTRY: dict[str, Callable] = {}
+"""A registry for different grid geometries.
 
-    This data class is used to setup the arrangement of grid cells to be used in
-    running a `virtual_rainforest` simulation.
+This dictionary maps grid geometry types (square, hex, etc) to a function generating a
+grid of that type. Users can register their own grid types using the `register_grid`
+decorator.
+"""
 
-    Attrs:
-        grid_type:
-        cell_contiguity:
-        cell_area:
-        cell_nx:
-        cell_ny:
-    """
 
-    grid_type: Literal["hex", "square"] = "square"
-    cell_contiguity: Literal["queen", "rook"] = "rook"
-    cell_area: float = 100
-    cell_nx: StrictInt = 10
-    cell_ny: StrictInt = 10
+def register_grid(grid_type: str):
+    """Add a grid type and creator function to the grid registry.
 
+    This decorator is used to add a function creating a grid layout to the registry of
+    accepted grids. The function should return a dictionary of shapely.geometry.Polygon
+    objects, keyed by a cell id str.
 
-class CoreGrid:
-    """Define the grid of cells used in a virtual rainforest simulation.
+    The grid_type argument is used to identify the grid creation function to be used by
+    the Grid class and in configuration files.
 
     Args:
-        config: A CoreGridConfig instance used to setup the grid
-
-    Attrs:
-        config: The CoreGridConfig used to create an instance
-        cell_dict: A dictionary, keyed by cell id, of dictionaries, containing the cell
-            polygon, centroid and neighbours for each cell.
+        grid_type: A name to be used to identify the grid creation function.
     """
 
-    def __init__(self, config: CoreGridConfig) -> None:
+    def decorator_register_grid(func: Callable):
+
+        # Add the grid type to the registry
+        if grid_type in GRID_REGISTRY:
+            LOGGER.warning(
+                "Grid type %s already exists and is being replaced", grid_type
+            )
+        GRID_REGISTRY[grid_type] = func
+
+        # Pass the function through for use, using @wraps to expose the original
+        # function dunder attributes (__name__, __doc__ etc.)
+        @wraps(func)
+        def wrapper_register_grid(*args, **kwargs):
+
+            return func(*args, **kwargs)
+
+        return wrapper_register_grid
```

Comment:
Why not just returning `func`? This wrapping is needed when you are doing something to the input arguments of `func` or its output, but you want to keep the dunder attributes, as you explain. But that's not the case here. in other words, I think you can safely do:
```suggestion
        return func
```

Reply:
Yeah - was overthinking it.